cmd/gb-vendor/vendor/imports.go

```diff
@@ -40,3 +43,28 @@ func ParseImports(root string) (map[string]bool, error) {
 	err := filepath.Walk(root, walkFn)
 	return pkgs, err
 }
+
+// FetchMetadata fetchs the remote metadata for path.
+func FetchMetadata(path string) (io.ReadCloser, error) {
+	url := fmt.Sprintf("https://%s?go-get=1", path)
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, nil
+}
+
+// ParseMetadata fetchs and decodes remote metadata for path.
+func ParseMetadata(path string) (string, string, string, error) {
+	rc, err := FetchMetadata(path)
+	if err != nil {
+		return "", "", "", err
+	}
+	defer rc.Close()
+
+	meta, err := parseMetaGoImports(rc)
+	if err != nil || len(meta) < 1 {
+		return "", "", "", err
+	}
+	return meta[0].Prefix, meta[0].VCS, meta[0].RepoRoot, nil
```

Comment:
Yeah. That's a fair comment, this code is pretty rough and ready. The
metaImport struct comes from discover.go, which I imported from
x/toosl/go/vcs, I dunno if I want to keep it, by parsing that xml is messy.

I'm also uncomfortable with passing around a Repository interface value
_and_ an extra bit of path, but I don't think I understand the problem well
enough to know the right data structure -- but hey, at least I got rid of
the MultiRepo type.

No matter how ugly, I want to get this into the hands of testers as soon as
possible, then refactor from there.

On Wed, Jun 3, 2015 at 8:10 PM, Dávid Kaya <notifications@github.com> wrote:

> In cmd/gb-vendor/vendor/imports.go
> <https://github.com/constabulary/gb/pull/183#discussion_r31609554>:
>
> > +	return resp.Body, nil
> > +}
> > +
> > +// ParseMetadata fetchs and decodes remote metadata for path.
> > +func ParseMetadata(path string) (string, string, string, error) {
> > +	rc, err := FetchMetadata(path)
> > +	if err != nil {
> > +		return "", "", "", err
> > +	}
> > +	defer rc.Close()
> > +
> > +	meta, err := parseMetaGoImports(rc)
> > +	if err != nil || len(meta) < 1 {
> > +		return "", "", "", err
> > +	}
> > +	return meta[0].Prefix, meta[0].VCS, meta[0].RepoRoot, nil
>
> Subjective opinion: Wouldn't it be "nicer" if we return struct with
> Prefix, VCS and RepoRoot attributes instead of 3 return values + error?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/constabulary/gb/pull/183/files#r31609554>.
>